modules/atomic-widgets/prop-type-migrations/path-resolver.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace Elementor\Modules\AtomicWidgets\PropTypeMigrations;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class Path_Resolver {
+	public static function resolve( string $pattern, array $data, bool $allow_missing_leaf = true ): array {
+		$segments = self::parse_path( $pattern );
+
+		return self::resolve_segments( $segments, $data, [], '', $allow_missing_leaf );
+	}
+
+	public static function resolve_with_wildcard_binding( string $pattern, array $data, array $wildcard_values ): ?string {
+		$segments = self::parse_path( $pattern );
+		$wildcard_index = 0;
+
+		return self::resolve_with_binding( $segments, $data, '', $wildcard_values, $wildcard_index );
+	}
+
+	public static function get( string $path, array $data ) {
+		$segments = self::parse_concrete_path( $path );
+		$current = $data;
+
+		foreach ( $segments as $segment ) {
+			if ( ! is_array( $current ) ) {
+				return null;
+			}
+
+			$key = $segment['value'];
+
+			if ( ! array_key_exists( $key, $current ) ) {
+				return null;
+			}
+
+			$current = $current[ $key ];
+		}
+
+		return $current;
+	}
+
+	private static function parse_path( string $path ): array {
+		$segments = [];
+		$current = '';
+		$length = strlen( $path );
+
+		for ( $i = 0; $i < $length; $i++ ) {
+			$char = $path[ $i ];
+
+			if ( '.' === $char ) {
+				if ( '' !== $current ) {
+					$segments[] = self::create_segment( $current, 'key' );
+					$current = '';
+				}
+			} elseif ( '[' === $char ) {
+				if ( '' !== $current ) {
+					$segments[] = self::create_segment( $current, 'key' );
+					$current = '';
+				}
+
+				$end = strpos( $path, ']', $i );
+
+				if ( false === $end ) {
+					break;
+				}
+
+				$index = substr( $path, $i + 1, $end - $i - 1 );
+				$segments[] = self::create_segment( $index, 'index' );
+				$i = $end;
+			} else {
+				$current .= $char;
+			}
+		}
+
+		if ( '' !== $current ) {
+			$segments[] = self::create_segment( $current, 'key' );
+		}
+
+		return $segments;
+	}
+
```

Comment:
I see this code block appears at least twice, can we extract it to a separate function? this way we also receive cleaner code by reducing nested conditions.

Maybe?

```suggestion
	private static function parse_path( string $path ): array {
	$segments = [];
	$current = '';
	$length = strlen( $path );

	for ( $i = 0; $i < $length; $i++ ) {
		$char = $path[ $i ];

		if ( '.' === $char ) {
			self::flush_current_segment( $segments, $current );
		} elseif ( '[' === $char ) {
			self::flush_current_segment( $segments, $current );

			$end = strpos( $path, ']', $i );

			if ( false === $end ) {
				break;
			}

			$index = substr( $path, $i + 1, $end - $i - 1 );
			$segments[] = self::create_segment( $index, 'index' );
			$i = $end;
		} else {
			$current .= $char;
		}
	}

	if ( '' !== $current ) {
		$segments[] = self::create_segment( $current, 'key' );
	}

	return $segments;
}

private static function flush_current_segment( array &$segments, string &$current ): void {
	if ( '' !== $current ) {
		$segments[] = self::create_segment( $current, 'key' );
		$current = '';
	}
}

```